logstash-core/src/main/java/org/logstash/plugins/pipeline/PipelineBus.java

```diff
@@ -58,26 +58,43 @@ public void sendEvents(final PipelineOutput sender,
             final ConcurrentHashMap<String, AddressState> addressesToInputs = outputsToAddressStates.get(sender);
 
             addressesToInputs.forEach((address, addressState) -> {
-                final Stream<JrubyEventExtLibrary.RubyEvent> clones = events.stream().map(e -> e.rubyClone(RubyUtil.RUBY));
-
-                PipelineInput input = addressState.getInput(); // Save on calls to getInput since it's volatile
-                boolean sendWasSuccess = input != null && input.internalReceive(clones);
-
-                // Retry send if the initial one failed
-                while (ensureDelivery && !sendWasSuccess) {
-                    // We need to refresh the input in case the mapping has updated between loops
-                    String message = String.format("Attempted to send event to '%s' but that address was unavailable. " +
-                            "Maybe the destination pipeline is down or stopping? Will Retry.", address);
-                    logger.warn(message);
-                    input = addressState.getInput();
-                    sendWasSuccess = input != null && input.internalReceive(clones);
-                    try {
-                        Thread.sleep(1000);
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        logger.error("Sleep unexpectedly interrupted in bus retry loop", e);
+                boolean sendWasSuccess = false;
+                PipelineInput.ReceiveResponse lastResponse = null;
+                boolean partialProcessing;
+                int lastFailedPosition = 0;
+                do {
+                    Stream<JrubyEventExtLibrary.RubyEvent> clones = events.stream()
+                            .skip(lastFailedPosition)
```

Comment:
a bit confusing but I guess iteration order would be always the same even if it was a `Set` (skipping the same elements).
likely always a `RubyArray` just typed loosely as `Collection<JrubyEventExtLibrary.RubyEvent> events` still looking at the `skip` initially made me assume it's at least a `List` ...

Reply:
so in both cases it is actually a `List` (which is itself created from iteration another list e.g. from the batch) :hammer: 
I think having another list (or a `orderedEvents = events.toArray()`), as you proposed, is acceptable here.

Reply:
Thank's nice suggestion. To keep the type of the array an empty typed array need to be created but should be lighter then creating a new `ArrayList`. Suggestion applied.